### PR TITLE
VK_EXT_shader_demote_to_helper_invocation

### PR DESCRIPTION
**Pull request by perlfu**
Implementation of VK_EXT_shader_demote_to_helper_invocation.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7283064c6569be30fbc707ed58b9b5738e0b75de
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

File: builder/llpcBuilderRecorder.cpp [hunk 1]
```diff
@@ -628,6 +632,21 @@ Value* BuilderRecorder::CreateDerivative(
     return Record(Opcode::Derivative, pValue->getType(), { pValue, getInt1(isDirectionY), getInt1(isFine) }, instName);
 }
 
+// Create a demote to helper invocation.
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: Miss a separate line // =====

File: builder/llpcBuilderImplMisc.cpp [hunk 2]
```diff
@@ -100,6 +100,27 @@ Instruction* BuilderImplMisc::CreateKill(
...
+
+    return CreateIntrinsic(Intrinsic::amdgcn_wqm_demote, {}, getFalse(), nullptr, instName);
+}
+
+// =====================================================================================================================
+// Create a helper invocation query. Only allowed in a fragment shader.
+Value* BuilderImplMisc::CreateIsHelperInvocation(
+    const Twine& instName) // [in] Name to give instruction(s)
+{
+    auto pIsHelper = CreateIntrinsic(Intrinsic::amdgcn_wqm_helper, {}, {}, nullptr, instName);
```
Comment (amdrexu) [inline builder/llpcBuilderImplMisc.cpp]: What is the return value of amdgcn_wqm_helper? pIsHelper means this is a helper lane? So why we still need a negation operation by CreateNot?

Comment (perlfu) [reply builder/llpcBuilderImplMisc.cpp]: wqm_helper is actually an alternate version of ps_live.  I agree the choice of name is misleading, it should probably be wqm_live, I'll change this in the upstream copy.

File: builder/llpcBuilderRecorder.cpp (hunk 1 above)
Comment (perlfu) [reply builder/llpcBuilderRecorder.cpp]: Fixed.

File: builder/llpcBuilderImplMisc.cpp (hunk 2 above)
Comment (amdrexu) [reply builder/llpcBuilderImplMisc.cpp]: OK. Then change to this: pIsHelper -> pIsNotHelper

**Comment by amdvlk-admin** [comment]
### Test summary for commit c23a9e3e586f46f06af5a4209d8ee37e93f6adc2
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0bb0956a8a17bebcf6217730e59f867f54248136
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 3b06acaf87d9028b65cdac6ebd394bc70637b573
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
Looks fine to me.

**Pull request merged**